### ISSUE TITLE
[v1.x] CI tests consistently failing in v1.x branch on onednn builds

### ISSUE THREAD

**Issue by josephevans** [apache/incubator-mxnet#20827]
## Description
The following tests seem to be failing consistently in the v1.x branch:

* test_autograd.test_retain_grad
  * ```terminate called after throwing an instance of 'dmlc::Error' what():  [04:06:05] src/imperative/imperative.cc:295: Check failed: !AGInfo::IsNone(*i): Cannot differentiate node because it is not in a computational graph. You need to set is_recording to true or use autograd.record() to save computational graphs for backward. If you want to differentiate the same graph twice, you need to pass retain_graph=True to backward.```
* test_gluon_gpu.test_sync_batchnorm
  * ```terminate called after throwing an instance of 'dmlc::Error' what():  [04:27:24] src/engine/threaded_engine.cc:333: Check failed: exec_ctx.dev_id < device_count_ (1 vs. 1) : Invalid GPU Id: 1, Valid device id should be less than device_count: 1```
* test_contrib_amp.test_amp_conversion_rnn
  * ```terminate called after throwing an instance of 'dmlc::Error' what():  [04:25:16] src/operator/./rnn-inl.h:508: RNN on GPU is only available for cuDNN at the moment.```

## Occurrences
* test_retain_grad:
  * https://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/PR-20809/8/pipeline/295
  * https://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-cpu/detail/PR-20809/8/pipeline/294
* test_contrib_amp.test_amp_conversion_rnn
  * https://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-20809/10/pipeline/355
* test_gluon_gpu.test_sync_batchnorm
  * https://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/mxnet-validation%2Funix-gpu/detail/PR-20809/10/pipeline/354

**Comment by anko-intel** [comment]
@josephevans, some of the fails from description are on MKLDNN-GPU. Do you know if it also happened on MKLDNN-CPU ?

**Comment by josephevans** [comment]
> @josephevans, some of the fails from description are on MKLDNN-GPU. Do you know if it also happened on MKLDNN-CPU ?

Hi @anko-intel, Unfortunately, this is happening in both unix-cpu and unix-gpu pipelines (see above links to Jenkins pipelines for examples for both.)

**Comment by anko-intel** [comment]
I have checked  the first example in the issue (test_autograd.test_retain_grad). 
It fails when the test is waiting for exception which should by throw by LogMessageFatal, but segfault occurs in destructor. It happens on environment where src/imperative/imperative.cc uses LogMessageFatal from the file 3rdparty/**mshadow/mshadow**/logging.h. 
On other than CI environment and when DNNL is not enabled, LogMessageFatal is taken from 3rdparty/**dmlc-core/i**nclude/dmlc/logging.h.
Unfortunately both files uses the same dmlc namespace. 
I have checked order of includes during compilation of imperative.cc and in both environment first dmlc version is included and later from mshadow. So I suppose that linking order take part in reproduction of the issue.
I have prepared short fix for it where I rename LogMessageFatal  in mshadow (https://github.com/apache/incubator-mxnet/pull/20838) , I think it is enough as temporary solution. 
As both logging.h files define macros with the same names (which use different implementation underneath) I think we should finally remove mshadow loggin version and use it from dmlc..
@szha  - any comments, any other ideas how to fix it?
